src/lib/TopLevel.hs

```diff
@@ -909,7 +936,57 @@ getFilteredLogger = do
   logger  <- getLogger
   return $ FilteredLogger shouldLog logger
 
--- -- === instances ===
+-- === hoisting Ix dicts ===
+
+hoistIxDicts
+  :: ( Topper m, Mut n, GenericallyTraversableE e
+     , SinkableE e, SubstE Name e, HoistableE e)
+  => e n -> m n (e n)
+hoistIxDicts e = do
+  Abs frag (PairE e' (IxHoistingState UnitE)) <-
+    liftGenericTraverserMTopEmissions (IxHoistingState UnitE) $ traverseGenericE (sink e)
+  e'' <- emitEnv (Abs frag e')
+  evalRequiredSpecializations e''  -- track the needed sepcializations rather than traversing again to find them
+  return e''
+
+newtype IxHoistingState (n::S) = IxHoistingState (UnitE n)
+        deriving (SinkableE, HoistableState, SubstE Name)
+
+instance GenericTraverser TopEnvFrag IxHoistingState where
+  traverseAtom = \case
+    DictCon d -> do
+      d' <- substM d
+      dTy <- getType d'
+      DictCon <$> case dTy of
+        DictTy dTy'@(DictType "Ix" _ _) -> emitIxDictSpecialization dTy' d'
+        _ -> return d'
+    atom -> traverseAtomDefault atom
+
+emitIxDictSpecialization :: HoistingTopBuilder m => DictType n -> DictExpr n -> m n (DictExpr n)
+emitIxDictSpecialization _ d@(ExplicitMethods _ _ _) = return d -- Nothing to do if we already have explicit methods
```

Comment:
Does that ever happen? Perhaps for top-level results that got inlined? We might consider adding a TODO to actually emit the dicts that top-level bound atoms refer to into the top-level too.

Reply:
Good question. I'll check. To be clear, `ExplicitMethods` doesn't inline the methods exactly. It refers versions of them by name, and supplies the local data they need.

Reply:
Looks like that code path does get hit in eval_tests. Is there a problem with it? I don't think I understand your suggestion about adding a TODO.

Reply:
Right, if you are 1000% sure that we never inline into `ExplicitMethods` then this code might be ok. But given that this is not checked statically, I am not sure. Since Ix emissions are temporary, what happens when they're using in a result that we'll have to put in the top level env? You won't be able to hoist the result atom above the Ix.

Reply:
Ok, I'll change it to be an error, and I'll also make `ExplicitMethods` take names instead of atoms to make the no-inlining guarantee more explicit. I think it should just require a custom (crashing) substitution rule.

Reply:
Oh, so what's you're saying is that this patch emits all of them permanently, which is not good, but it will be ok once the other patch lands?

Reply:
Yes, that's right.